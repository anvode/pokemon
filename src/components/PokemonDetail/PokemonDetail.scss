.pokemon-detail {
    &__shrink {
        white-space: nowrap;
        width: 1px;
    }
}
