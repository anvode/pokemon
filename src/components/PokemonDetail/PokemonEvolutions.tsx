import * as React from 'react';
import Loader from '../Loader/Loader';
import { useFetchEvolutionChain } from './pokemon.utils';
import { capitalize } from '../../Helpers';

export interface PokemonEvolutionsProps {
    species: {
        name: string;
        url: string;
    };
}
const PokemonEvolutions: React.FC<PokemonEvolutionsProps> = ({ species }) => {
    const index = species.url.split('/')[species.url.split('/').length - 2];

    const {loading, error, chains} = useFetchEvolutionChain(+index);
    
    if (error) {
        return <div>
            Oops! An Error happens while Fetching:  <strong>{error}</strong>
        </div>;
    }

    if (loading) {
        return <Loader></Loader>;
    }
    
    return <>
        <div className="card-header border-0 rounded-0 text-white bg-primary mt-3">
            <h2 className="card-title h4 mb-0">Evolution Chain</h2>
        </div>
        <div className="d-flex mb-0 border border-top-0 p-3 flex-wrap">
            {chains && chains.map((chain, index) => {
                return <React.Fragment key={chain.species_name}>
                    {index !== 0 && (
                        <div className="pokemon-evolution d-flex justify-content-around flex-grow-1">
                            <div className="px-2 text-center">
                                {chain.min_level && <>
                                    <strong>Level</strong> <br/>
                                    {chain.min_level} <br/>
                                </>}
                                â†’
                            </div>
                        </div>
                        
                    )}
                    <div className="pokemon-evolution d-flex align-items-center justify-content-around flex-grow-1">
                        <div className="px-2 text-primary h4">
                            {capitalize(chain.species_name)}
                        </div>
                    </div>
                  
                </React.Fragment>;
            })}
        </div>
    </>; 
};

export default PokemonEvolutions;