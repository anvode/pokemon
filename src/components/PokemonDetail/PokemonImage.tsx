import * as React from 'react';
import { useEffect, useState } from 'react';
import placeholder from '../../assets/not.png';

export interface PokemonImageProps {
    name: string;
    url: string | null;
}
const PokemonImage: React.FC<PokemonImageProps> = ({name, url}) => {
    const [imageLoading, setImageLoading] = useState(true);
    const [imageUrl, setImageUrl] = useState('');

    useEffect(() => {
        const image = url ? url : placeholder;
        setImageUrl(image);
    }, []);

    return <>
        <div className="pokemon-detail__image h-100 border align-items-center d-flex justify-content-center">
            <img
                className="img-fluid"
                src={imageUrl}
                alt={name}
                onLoad={() => setImageLoading(false)}
                onError={() => {
                    setImageLoading(false);
                    setImageUrl(placeholder);
                }}
                style={
                    imageLoading
                        ? { display: 'none' }
                        : { display: 'block' }
                }
            />
        </div>
    </>; 
};

export default PokemonImage;