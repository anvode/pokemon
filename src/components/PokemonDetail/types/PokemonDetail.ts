
export interface PokemonDetailState {
    pokemonDetailLoading: boolean;
    pokemonDetailError: boolean;
    pokemonDetail: Pokemon | null;
    id: number | null
}

export interface Filter {
    [key: string]: string;
}

export interface Pokemon {
    sprites: Image;
    name: string;
    order: number;
    abilities: Ability[];
    types: Type[];
    stats: State[];
    species: {
        name: string;
        url: string;
    };
    moves: Move[];
    [key: string]: number | string | boolean | object | null | undefined;
}

export interface Image {
    [key: string]: string | null;
}

export interface Move {
    name: string;
    url: string;
}

export interface Ability {
    ability: {
        name: string;
        url: string;
    };
    is_hidden: boolean;
    slot: number;
}

export interface Venue {
    city: string;
    houseNumber: string;
    name: string;
    street: string;
    uid: number;
    zipCode: string;
}

export interface Type {
    slot: number;
    type: {
        name: string;
        url: string;
    };
}

export interface State {
    base_stat: number;
    effort: number;
    stat: {
        name: string;
        url: string;
    };
}

export interface Evolution {
    species_name: string,
    min_level: number,
    trigger_name: string,
    url: string | null
}

