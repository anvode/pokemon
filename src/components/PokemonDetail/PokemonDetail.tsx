
import * as React from 'react';
import { useEffect, useReducer } from 'react';
import { useParams } from 'react-router-dom';

import { PokemonDetailState } from './types/PokemonDetail';
import './PokemonDetail.scss';
import Title from '../Title/Title';
import { pokemonDetailReducer } from './pokemon.reducer';
import { fetchPokemon } from './pokemon.utils';
import Loader from '../Loader/Loader';
import { capitalize } from '../../Helpers';
import PokemonProperties from './PokemonProperties';
import PokemonImage from './PokemonImage';
import PokemonStats from './PokemonStats';
import PokemonMoves from './PokemonMoves';
import PokemonEvolutions from './PokemonEvolutions';

export interface PokemonDetailProps {}

export const initialState: PokemonDetailState = {
    pokemonDetailLoading: false,
    pokemonDetailError: false,
    pokemonDetail: null,
    id: null
};

const PokemonDetail: React.FC<PokemonDetailProps> = (props) => {
    const [state, dispatch] = useReducer(pokemonDetailReducer, initialState);
    const {id} = useParams();

    useEffect(() => {
        fetchPokemon(dispatch, id);

    }, []);

    const {pokemonDetailLoading, pokemonDetailError, pokemonDetail} = state;
    
    if (pokemonDetailError) {
        return <div>
            Oops! An Error happens while Fetching:  <strong>{pokemonDetailError}</strong>
        </div>;
    }

    if (pokemonDetailLoading || pokemonDetail === null) {
        return <Loader></Loader>;
    }
    
    const { name, sprites, order, abilities, types, moves, stats, species} = pokemonDetail;

    return <>
        <Title name={`${capitalize(name)}`}></Title>
        <div className="row mt-1">
            <div className="col-md-4 mb-3">
                <PokemonImage name={name} url={sprites.front_default}></PokemonImage>
            </div>
            <div className="col-md-8 mb-3">
                <PokemonProperties name={name} order={order} abilities={abilities} types={types}></PokemonProperties>
            </div>
        </div>
        <PokemonStats stats={stats}></PokemonStats>
        <PokemonEvolutions species={species}></PokemonEvolutions>
        <PokemonMoves moves={moves}></PokemonMoves>
    </>;
};

export default PokemonDetail;