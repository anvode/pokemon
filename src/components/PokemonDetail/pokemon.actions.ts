import { Pokemon } from './types/PokemonDetail';

import { POKEMON_DETAIL_LIST, POKEMON_DETAIL_LOADING, POKEMON_DETAIL_ERROR, PokemonDetailActionTypes } from './types/actions';

export const SetPokemonDetailAction = (pokemonDetail: Pokemon, id: number): PokemonDetailActionTypes => {
    return {
        type: POKEMON_DETAIL_LIST,
        payload: {
            pokemonDetail,
            id
        }
    };
};

export const setPokemonDetailErrorAction = (error: boolean): PokemonDetailActionTypes => {
    return {
        type: POKEMON_DETAIL_ERROR,
        payload: {
            error
        }
    };
};

export const setPokemonDetailLoadingAction = (loading: boolean): PokemonDetailActionTypes => {
    return {
        type: POKEMON_DETAIL_LOADING,
        payload: {
            loading
        }
    };
};
