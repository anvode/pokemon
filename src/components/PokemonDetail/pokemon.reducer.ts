import { PokemonDetailState } from './types/PokemonDetail';

import { POKEMON_DETAIL_LIST, POKEMON_DETAIL_LOADING, POKEMON_DETAIL_ERROR, PokemonDetailActionTypes } from './types/actions';

export const pokemonDetailReducer = (state: PokemonDetailState, action: PokemonDetailActionTypes): PokemonDetailState => {

    switch (action.type) {

        case POKEMON_DETAIL_LIST: {
            return {
                ...state,
                pokemonDetail: action.payload.pokemonDetail,
                id: action.payload.id
            };
        }

        case POKEMON_DETAIL_ERROR: {
            return {
                ...state,
                pokemonDetailError: action.payload.error
            };
        }

        case POKEMON_DETAIL_LOADING: {
            return {
                ...state,
                pokemonDetailLoading: action.payload.loading
            };
        }

        default:
            return state;
    }
};
