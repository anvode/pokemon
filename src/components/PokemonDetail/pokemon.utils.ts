import { useEffect, useState } from 'react';

import { PokemonDetailActionTypes } from './types/actions';
import { setPokemonDetailErrorAction, setPokemonDetailLoadingAction, SetPokemonDetailAction } from './pokemon.actions';
import { Move, Evolution } from './types/PokemonDetail';

export const fetchPokemon = async (dispatch: React.Dispatch<PokemonDetailActionTypes>, id: number) => {
 
    dispatch(setPokemonDetailLoadingAction(true));
    try {
        const response = await fetch(`${process.env.REACT_APP_POKEMON_API_URL}/${id}`);

        const {
            sprites, name, order, abilities, types, stats, species, moves: filterMoves
        } = await response.json();
       
        const moves: Move[] = filterMoves.map((item: {
            move: Move;
        }) => item.move);

        dispatch(setPokemonDetailLoadingAction(false));
        dispatch(SetPokemonDetailAction({sprites, name, order, abilities, types, stats, species, moves}, id));

    } catch (err) {
        dispatch(setPokemonDetailLoadingAction(false));
        dispatch(setPokemonDetailErrorAction(err.message));
    }
};

export const useFetchEvolutionChain = (id: number) => {
 
    const [chains, setChain] = useState<Evolution[] | null>(null);
    const [error, setError] = useState(false);
    const [loading, setLoading] = useState(false);

    useEffect(() => {
        const fetchData = async () => {
            setLoading(true);

            try {
                const responseSpecies = await fetch(`${process.env.REACT_APP_POKEMON_SPECIES_API_URL}/${id}`);
                const {evolution_chain} = await responseSpecies.json();

                const responseChain = await fetch(`${evolution_chain.url}`);
                const dataChain = await responseChain.json();   

                const getEvolutionChain = setEvolutionChain(dataChain.chain);
                
                setLoading(false);
                setChain(getEvolutionChain);

            } catch (err) {
                setLoading(false);
                setError(err.message);
            }
        };

        fetchData();

    }, [id]);

    return {
        loading, error, chains
    };

};

const setEvolutionChain = (chain: any): Evolution[] | null => {
    const evoChain = [];
    let evoData = chain;
    
    do {
        const evoDetails = evoData['evolution_details'][0];
    
        evoChain.push({
            'species_name': evoData.species.name,
            'min_level': !evoDetails ? 1 : evoDetails.min_level,
            'trigger_name': !evoDetails ? null : evoDetails.trigger.name,
            'url': !evoDetails ? null : evoData.species.url
        });
    
        evoData = evoData['evolves_to'][0];

    } while (!!evoData && Object.prototype.hasOwnProperty.call(evoData, 'evolves_to'));

    return evoChain.length ? evoChain : null;
};