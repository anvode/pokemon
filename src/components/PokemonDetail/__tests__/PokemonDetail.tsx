import * as React from 'react';
import { act } from 'react-dom/test-utils';
import { render, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';

import Store from '../../../redux/store';
import PokemonDetail from '../PokemonDetail';

describe('PokemonDetail', () => {
    afterEach(cleanup);

    it('renders without crashing', async () => {

        await act(async () => {
            const { asFragment } = render(<Provider store={Store} ><BrowserRouter><PokemonDetail /></BrowserRouter></Provider>);
            expect(asFragment()).toMatchSnapshot(); });
       
    });
});

