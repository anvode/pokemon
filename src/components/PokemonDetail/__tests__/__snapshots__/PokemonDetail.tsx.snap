// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonDetail renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="loader__container"
  >
    <div
      class="loader"
      data-testid="loader"
    >
      <div />
      <div />
      <div />
      <div />
    </div>
  </div>
</DocumentFragment>
`;
